### PR TITLE
Add 'refresh' mode to fact gathering

### PR DESCRIPTION
**Pull request by electrofelix**
##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
executor/play_iterator

##### ANSIBLE VERSION
```
ansible 2.3.0
  config file = /home/baileybd/.ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

Ensures that gathered facts are refreshed at the start of play
execution, still using the cache for all remaining plays same as
'smart'.

Most useful when combined with a persistent fact cache store such as
'jsonfile' or 'redis', where you may wish to persist custom facts
stored in addition that you don't wish to flush or expire as quickly as
system facts gathered.

This allows for refreshing system facts at the start of a playbook,
and then continuing to benefit from the 'smart' behaviour during the
subsequent plays executed as part of that play.

test-gathering.yml

```
- hosts: localhost
  tasks:
    - ping:
      register: ping_status

- hosts: localhost
  tasks:
    - debug:
        var: ping_status
```

ansible.cfg

```
[defaults]
gathering = smart
fact_caching = jsonfile
fact_caching_connection = ~/.ansible_fact_cache
fact_caching_timeout = 86400
```

Run 1

```
ansible-playbook test-gathering.yml 

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [ping] ********************************************************************
ok: [localhost]

PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ping_status": {
        "changed": false, 
        "ping": "pong"
    }
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

Run 2

```
ansible-playbook test-gathering.yml 

PLAY [localhost] ***************************************************************

TASK [ping] ********************************************************************
ok: [localhost]

PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ping_status": {
        "changed": false, 
        "ping": "pong"
    }
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

Notice that for the second run in addition to skipping gather on the second play, it also skipped it on the first as well.

ansible.cfg

```
[defaults]
gathering = refresh
fact_caching = jsonfile
fact_caching_connection = ~/.ansible_fact_cache
fact_caching_timeout = 86400
```

Run 1

```
ansible-playbook test-gathering.yml 

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [ping] ********************************************************************
ok: [localhost]

PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ping_status": {
        "changed": false, 
        "ping": "pong"
    }
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

Run 2

```
ansible-playbook test-gathering.yml 

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [ping] ********************************************************************
ok: [localhost]

PLAY [localhost] ***************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "ping_status": {
        "changed": false, 
        "ping": "pong"
    }
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```

With _refresh_, it will re-gather for the first play in each run, but then use the fact cache for all remaining plays. This also allows us to ensure we always have a consistent view of basic system 
state at the start of a playbook run when using fact caching that can contain additional custom 
facts set from previous runs, while benefiting from caching for the remaining plays.

### CODE REVIEW

**Comment by electrofelix** [comment]
I originally did this for 1.9, where custom facts would be stored in the same fact cache, so if you used memory backing fact cache you would lose custom facts that had been returned by any module from a previous run. Forcing us to add calls to force regathering of system facts on subsequent runs where we wanted to ensure that we had up to date basic system information. Adding a refresh behaviour for fact gathering allow us to persist up to the cache timeout the custom facts from some modules returned, while always expiring at the start of a run the system facts and have them regathered when using a persistent fact cache.

Using this allowed for refresh mode to be set in the config file along with a persistent fact cache and take actions on previously set custom facts if run within a certain time.

With all the work done around custom facts, it maybe that a better pattern exists now, and this is no longer required.

**Review by dharmabumstead — APPROVED** [review]

**Comment by dharmabumstead** [comment]
Looks OK from a documentation standpoint. Still needs tech review from a core dev. Thanks.

**Comment by calfonso** [comment]
Hi!

Thanks very much for your submission to Ansible. It sincerely means a lot to us that you've taken time to contribute.

- `meta: clear_facts` was added to accomplish the same thing

However, we're absolutely always up for discussion. Since this is a really busy project, we don't always see comments on closed tickets, but want to encourage open dialog. You can stop by the development list, and we'd be glad to talk about it - and we might even be persuaded otherwise!

https://groups.google.com/forum/#!forum/ansible-devel
In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

**Pull request closed without merging**
